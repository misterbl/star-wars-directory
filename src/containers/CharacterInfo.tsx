import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { Formik } from "formik";
import * as Yup from "yup";
import { bindActionCreators, Dispatch } from "redux";
import {
  getInfo,
  sendingInfoSms,
  getSpecies,
  getHomeWorld,
  getCharacterFilms,
  getVehicles,
  getStarships
} from "../selectors/appSelectors";
import { IAppState } from "../state";
import { ICharacterInfoComponent } from "./CharacterInfo.d";
import ROUTES from "../const/routes";
import charactersPhotos from "../const/charactersPhotos";
import BackButton from "../components/BackButton";
import SendInfoSmsForm from "../components/SendInfoSmsForm";
import sendInfoSms from "../actions/thunks/sendInfoSms";
import { phoneRegExp } from "../const/regex";
import InfoCategory from "../components/InfoCategory";
export class CharacterInfo extends Component<ICharacterInfoComponent> {
  state = {
    displaySentSuccess: false
  };
  componentWillReceiveProps() {
    const {
      character: { name },
      history: { push }
    } = this.props;
    if (!name) {
      push(ROUTES.INDEX);
    }
  }
  handleSubmit = (e: any) => {
    const {
      sendInfoSms,
      character: { name, height, hair_color, eye_color, birth_year, gender }
    } = this.props;
    const number = e.phoneNumber;
    const message = `Here is your search result from Star Wars: ${name}, ${gender}, born in ${birth_year}, mass: ${height}cm, ${hair_color} hair, ${eye_color} eyes`;
    sendInfoSms(message, number);
    this.setState({ displaySentSuccess: true });
  };

  resetDisplaySentSuccess = () => {
    this.setState({ displaySentSuccess: false });
  };

  render() {
    const {
      character,
      history: { push },
      sendingInfoSms,
      species,
      homeworld,
      characterFilms,
      vehicles,
      starShips
    } = this.props;
    const { displaySentSuccess } = this.state;

    const sprite = charactersPhotos[character.name];
    const { name, height, gender } = character;

    return (
      <>
        {name && (
          <main className="p-5">
            <BackButton
              route={ROUTES.INDEX}
              text="Back to search result"
              push={push}
            />
            <Formik
              validationSchema={Yup.object().shape({
                phoneNumber: Yup.string()
                  .required("Phone number required")
                  .length(11)
                  .matches(phoneRegExp, "Phone number is not valid")
              })}
              initialValues={{
                phoneNumber: ""
              }}
              onSubmit={this.handleSubmit}
              render={formikProps => (
                <SendInfoSmsForm
                  sendingInfoSms={sendingInfoSms}
                  displaySentSuccess={displaySentSuccess}
                  resetDisplaySentSuccess={this.resetDisplaySentSuccess}
                  {...formikProps}
                />
              )}
            />
            <h2 className="text-white ml-5 mt-5 p-3">{name.toUpperCase()}</h2>
            {sprite ? (
              <div className="ml-5 avatar my-2">
                <div
                  style={{
                    backgroundPosition: `${sprite}`,
                    zoom: 2
                  }}
                />
              </div>
            ) : (
              <div className="avatar avatar__default mr-3" />
            )}
            <div className="d-flex p-4">
              <InfoCategory list={characterFilms} title="APPARENCIES" />
              <InfoCategory list={[homeworld]} title="HOMEWORLD" />
              <InfoCategory list={[gender]} title="GENDER" />
              <InfoCategory list={[`Height: ${height}cm`]} title="DIMENSION" />
              <InfoCategory list={[species]} title="SPECIES" />
              <InfoCategory list={vehicles} title="VEHICLES" />
              <InfoCategory list={starShips} title="STARSHIPS" />
            </div>
          </main>
        )}
      </>
    );
  }
}

export const mapStateToProps = (state: IAppState) => ({
  character: getInfo(state),
  sendingInfoSms: sendingInfoSms(state),
  species: getSpecies(state),
  homeworld: getHomeWorld(state),
  characterFilms: getCharacterFilms(state),
  vehicles: getVehicles(state),
  starShips: getStarships(state)
});

export const mapDispatchToProps = (dispatch: Dispatch) => ({
  sendInfoSms: bindActionCreators(sendInfoSms, dispatch)
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(CharacterInfo)
);
