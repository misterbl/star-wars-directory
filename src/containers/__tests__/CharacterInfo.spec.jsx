/* eslint-disable no-undef */
import * as React from "react";
import { shallow } from "enzyme";
import history from "../../testMocks/history.mock";
import { CharacterInfo } from "../CharacterInfo";
import ROUTES from "../../const/routes";

describe("CharacterInfo", () => {
  const props = {
    history,
    character: {
      name: "name",
      height: "123",
      mass: "12",
      birth_year: "1234",
      gender: "gender",
      hair_color: "hair_color",
      eye_color: "eye_color"
    }
  };
  const wrapper = shallow(<CharacterInfo {...props} />);
  it("matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("redirect to index if there is no character", () => {
    wrapper.setProps({ character: null });
    expect(props.history.push).toHaveBeenCalledWith(ROUTES.INDEX);
  });
});
