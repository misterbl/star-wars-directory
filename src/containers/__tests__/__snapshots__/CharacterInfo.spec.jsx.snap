// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterInfo matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CharacterInfo
    character={
      Object {
        "birth_year": "1234",
        "eye_color": "eye_color",
        "gender": "gender",
        "hair_color": "hair_color",
        "height": "123",
        "mass": "12",
        "name": "name",
      }
    }
    history={
      Object {
        "action": "PUSH",
        "block": [MockFunction],
        "createHref": [MockFunction],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [MockFunction],
        "length": 123,
        "listen": [MockFunction],
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": "",
        },
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <BackButton
          push={[MockFunction]}
          route="/"
          text="Back to search result"
        />,
        <div
          className="d-flex ml-5 mt-5 pl-5"
        >
          <div
            className="avatar avatar__default card-img-top mr-3"
          />
          <div
            className="info-details p-3"
          >
            <h2
              className="text-center"
            >
              NAME
            </h2>
            <p>
              Height: 123cm
            </p>
            <p>
              Weight: 12kgs
            </p>
            <p>
              Hair color: hair_color
            </p>
            <p>
              Skin color: undefined
            </p>
            <p>
              Eye color: eye_color
            </p>
            <p>
              Born in 1234
            </p>
            <p>
              Gender: gender
            </p>
          </div>
        </div>,
        <Formik
          enableReinitialize={false}
          initialValues={
            Object {
              "phoneNumber": "",
            }
          }
          isInitialValid={false}
          onSubmit={[Function]}
          render={[Function]}
          validateOnBlur={true}
          validateOnChange={true}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "push": [MockFunction],
          "route": "/",
          "text": "Back to search result",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="avatar avatar__default card-img-top mr-3"
            />,
            <div
              className="info-details p-3"
            >
              <h2
                className="text-center"
              >
                NAME
              </h2>
              <p>
                Height: 123cm
              </p>
              <p>
                Weight: 12kgs
              </p>
              <p>
                Hair color: hair_color
              </p>
              <p>
                Skin color: undefined
              </p>
              <p>
                Eye color: eye_color
              </p>
              <p>
                Born in 1234
              </p>
              <p>
                Gender: gender
              </p>
            </div>,
          ],
          "className": "d-flex ml-5 mt-5 pl-5",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "avatar avatar__default card-img-top mr-3",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h2
                  className="text-center"
                >
                  NAME
                </h2>,
                <p>
                  Height: 123cm
                </p>,
                <p>
                  Weight: 12kgs
                </p>,
                <p>
                  Hair color: hair_color
                </p>,
                <p>
                  Skin color: undefined
                </p>,
                <p>
                  Eye color: eye_color
                </p>,
                <p>
                  Born in 1234
                </p>,
                <p>
                  Gender: gender
                </p>,
              ],
              "className": "info-details p-3",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "NAME",
                  "className": "text-center",
                },
                "ref": null,
                "rendered": "NAME",
                "type": "h2",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Height: 123cm",
                },
                "ref": null,
                "rendered": "Height: 123cm",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Weight: 12kgs",
                },
                "ref": null,
                "rendered": "Weight: 12kgs",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Hair color: hair_color",
                },
                "ref": null,
                "rendered": "Hair color: hair_color",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Skin color: undefined",
                },
                "ref": null,
                "rendered": "Skin color: undefined",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Eye color: eye_color",
                },
                "ref": null,
                "rendered": "Eye color: eye_color",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Born in 1234",
                },
                "ref": null,
                "rendered": "Born in 1234",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Gender: gender",
                },
                "ref": null,
                "rendered": "Gender: gender",
                "type": "p",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "enableReinitialize": false,
          "initialValues": Object {
            "phoneNumber": "",
          },
          "isInitialValid": false,
          "onSubmit": [Function],
          "render": [Function],
          "validateOnBlur": true,
          "validateOnChange": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <BackButton
            push={[MockFunction]}
            route="/"
            text="Back to search result"
          />,
          <div
            className="d-flex ml-5 mt-5 pl-5"
          >
            <div
              className="avatar avatar__default card-img-top mr-3"
            />
            <div
              className="info-details p-3"
            >
              <h2
                className="text-center"
              >
                NAME
              </h2>
              <p>
                Height: 123cm
              </p>
              <p>
                Weight: 12kgs
              </p>
              <p>
                Hair color: hair_color
              </p>
              <p>
                Skin color: undefined
              </p>
              <p>
                Eye color: eye_color
              </p>
              <p>
                Born in 1234
              </p>
              <p>
                Gender: gender
              </p>
            </div>
          </div>,
          <Formik
            enableReinitialize={false}
            initialValues={
              Object {
                "phoneNumber": "",
              }
            }
            isInitialValid={false}
            onSubmit={[Function]}
            render={[Function]}
            validateOnBlur={true}
            validateOnChange={true}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "push": [MockFunction],
            "route": "/",
            "text": "Back to search result",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="avatar avatar__default card-img-top mr-3"
              />,
              <div
                className="info-details p-3"
              >
                <h2
                  className="text-center"
                >
                  NAME
                </h2>
                <p>
                  Height: 123cm
                </p>
                <p>
                  Weight: 12kgs
                </p>
                <p>
                  Hair color: hair_color
                </p>
                <p>
                  Skin color: undefined
                </p>
                <p>
                  Eye color: eye_color
                </p>
                <p>
                  Born in 1234
                </p>
                <p>
                  Gender: gender
                </p>
              </div>,
            ],
            "className": "d-flex ml-5 mt-5 pl-5",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "avatar avatar__default card-img-top mr-3",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <h2
                    className="text-center"
                  >
                    NAME
                  </h2>,
                  <p>
                    Height: 123cm
                  </p>,
                  <p>
                    Weight: 12kgs
                  </p>,
                  <p>
                    Hair color: hair_color
                  </p>,
                  <p>
                    Skin color: undefined
                  </p>,
                  <p>
                    Eye color: eye_color
                  </p>,
                  <p>
                    Born in 1234
                  </p>,
                  <p>
                    Gender: gender
                  </p>,
                ],
                "className": "info-details p-3",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "NAME",
                    "className": "text-center",
                  },
                  "ref": null,
                  "rendered": "NAME",
                  "type": "h2",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Height: 123cm",
                  },
                  "ref": null,
                  "rendered": "Height: 123cm",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Weight: 12kgs",
                  },
                  "ref": null,
                  "rendered": "Weight: 12kgs",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Hair color: hair_color",
                  },
                  "ref": null,
                  "rendered": "Hair color: hair_color",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Skin color: undefined",
                  },
                  "ref": null,
                  "rendered": "Skin color: undefined",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Eye color: eye_color",
                  },
                  "ref": null,
                  "rendered": "Eye color: eye_color",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Born in 1234",
                  },
                  "ref": null,
                  "rendered": "Born in 1234",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Gender: gender",
                  },
                  "ref": null,
                  "rendered": "Gender: gender",
                  "type": "p",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "enableReinitialize": false,
            "initialValues": Object {
              "phoneNumber": "",
            },
            "isInitialValid": false,
            "onSubmit": [Function],
            "render": [Function],
            "validateOnBlur": true,
            "validateOnChange": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
